sync2/rangesync/rangesync.go

```diff
@@ -0,0 +1,680 @@
+type Tracer interface {
+	// OnDumbSync is called when the difference metric exceeds maxDiff and dumb
+	// reconciliation process is used
+	OnDumbSync()
+	// OnRecent is invoked when Recent message is received
+	OnRecent(receivedItems, sentItems int)
+}
+
+type nullTracer struct{}
+
+func (t nullTracer) OnDumbSync()       {}
+func (t nullTracer) OnRecent(int, int) {}
+
+// WithTracer specifies a tracer for RangeSetReconciler.
+func WithTracer(t Tracer) RangeSetReconcilerOption {
+	return func(r *RangeSetReconciler) {
+		r.tracer = t
+	}
+}
+
+// WithClock specifies the clock for RangeSetReconciler.
+func WithClock(c clockwork.Clock) RangeSetReconcilerOption {
+	return func(r *RangeSetReconciler) {
+		r.clock = c
+	}
+}
+
+// ProbeResult contains the result of a probe.
+type ProbeResult struct {
+	// Fingerprint of the range.
+	FP any
+	// Number of items in the range.
+	Count int
+	// An estimate of Jaccard similarity coefficient between the sets.
+	// The range is 0..1, 0 being mostly disjoint sets and 1 being mostly equal sets.
+	Sim float64
+}
+
+// RangeSetReconciler reconciles two sets of items using the recursive set reconciliation
+// protocol.
+type RangeSetReconciler struct {
+	os             OrderedSet
+	maxSendRange   int
+	itemChunkSize  int
+	sampleSize     int
+	maxDiff        float64
+	recentTimeSpan time.Duration
+	tracer         Tracer
+	clock          clockwork.Clock
+	log            *zap.Logger
+}
+
+// NewRangeSetReconciler creates a new RangeSetReconciler.
+func NewRangeSetReconciler(os OrderedSet, opts ...RangeSetReconcilerOption) *RangeSetReconciler {
+	rsr := &RangeSetReconciler{
+		os:            os,
+		maxSendRange:  DefaultMaxSendRange,
+		itemChunkSize: DefaultItemChunkSize,
+		sampleSize:    DefaultSampleSize,
+		maxDiff:       -1,
+		tracer:        nullTracer{},
+		clock:         clockwork.NewRealClock(),
+		log:           zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(rsr)
+	}
+	if rsr.maxSendRange <= 0 {
+		panic("bad maxSendRange")
+	}
+	return rsr
+}
+
+func (rsr *RangeSetReconciler) defaultRange() (x, y KeyBytes, err error) {
+	if empty, err := rsr.os.Empty(); err != nil {
+		return nil, nil, fmt.Errorf("checking for empty set: %w", err)
+	} else if empty {
+		return nil, nil, nil
+	}
+
+	x, err = rsr.os.Items().First()
+	if err != nil {
+		return nil, nil, fmt.Errorf("get items: %w", err)
+	}
+
+	return x, x, nil
+}
+
+func (rsr *RangeSetReconciler) processSubrange(s sender, x, y KeyBytes, info RangeInfo) error {
+	rsr.log.Debug("processSubrange", log.ZShortStringer("x", x), log.ZShortStringer("y", y),
+		zap.Int("count", info.Count), log.ZShortStringer("fingerprint", info.Fingerprint))
+
+	if info.Count == 0 {
+		// We have no more items in this subrange.
+		// Ask peer to send any items it has in the range
+		rsr.log.Debug("processSubrange: send empty range", log.ZShortStringer("x", x), log.ZShortStringer("y", y))
+		if err := s.SendEmptyRange(x, y); err != nil {
+			return err
```

Comment:
from this point onwards missing error instrumentation in multiple places

Reply:
Fixed